### ISSUE TITLE
TestTaskRunTimeout is flakey : fix timeouts

### ISSUE THREAD

**Issue by vdemeester** [tektoncd/pipeline#731]
# Expected Behavior

`TestTaskRunTimeout` should pass all the time :100: 

# Actual Behavior

`TestTaskRunTimeout` is failing a lot. I'll skip it so that 

Here are some examples of failures:
* From https://github.com/tektoncd/pipeline/pull/713:
  * https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/713/pull-tekton-pipeline-integration-tests/1113345177981816832/
  * https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/713/pull-tekton-pipeline-integration-tests/1113137811692720130/
  * https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/713/pull-tekton-pipeline-integration-tests/1113331336749256704/
* From https://github.com/tektoncd/pipeline/pull/711:
  * https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/711/pull-tekton-pipeline-integration-tests/1113053757723119616/
* From https://github.com/tektoncd/pipeline/pull/691:
  * https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/691/pull-tekton-pipeline-integration-tests/1116632841082376194/

# Steps to Reproduce the Problem

1. Open a PR
2. See it fail :hankey: 
3. Run `/retest` and :pray: 

I'm working on it in https://github.com/tektoncd/pipeline/pull/691 but it's at a point where it makes a lot of build fail, and I'm really thinking of marking the test as skip and track the re-enablement of it here

cc @bobcatfish 

/kind bug
/assign

**Comment by bobcatfish** [comment]
@vdemeester just let me know if you want a break from the investigation for a while, I'd be happy to take it on for a while :dancer:

**Comment by vdemeester** [comment]
/assign @bobcatfish 

The more the merrier (especially as I'm running out of time this week)

**Comment by bobcatfish** [comment]
tl;dr: The timeout handle and the reconciler disagree about when a Run has timed out.

Okay I have a theory about what's going on. In https://github.com/tektoncd/pipeline/pull/771 I ran the timeout test 10 times (and p.s. this _seems_ to reproduce the issue locally as well), and in https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/771/pull-tekton-pipeline-integration-tests/1118981945099816962/ we have 2/10 of those timing out.

I also added [a lot of (terrible) logs](https://github.com/tektoncd/pipeline/pull/771/commits/c08abb84bafc0478b30e00066704555592c49344) so I could see what was going on and I noticed some interesting things:

- The failed Runs _were_ considered timed out in [the timeout handler](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/timeout_handler.go) (i.e. [`case <-time.After(timeout)`](https://github.com/tektoncd/pipeline/blob/984049e315afc7dd5c381a87c3afe199906a18b0/pkg/reconciler/timeout_handler.go#L192) occurred) for all 10 Runs (and one of them twice haha)
- But for `run-giraffe-0` and `run-giraffe-6`, the subsequent reconcile did **NOT** consider them timed out

For example for `run-giraffe-0`, this is from the timeout handler:

```
I0418 21:17:16.127] {"level":"info","logger":"controller.taskrun-controller","caller":"reconciler/timeout_handler.go:203","msg":"TIMEOUT timed out for run-giraffe-0","knative.dev/controller":"taskrun-controller"}                    
I0418 21:17:16.127] {"level":"info","logger":"controller.taskrun-controller","caller":"reconciler/timeout_handler.go:200","msg":"TIMEOUT finished for run-giraffe-0","knative.dev/controller":"taskrun-controller"}
```

This would then trigger the reconcile loop, and the logs from that look like:

```
I0418 21:17:16.128] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:248","msg":"TIMEOUT about to check if timeout for run-giraffe-0","knative.dev/controller":"taskrun-controller"}               
I0418 21:17:16.128] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:531","msg":"TIMEOUT checking timeout for run-giraffe-0, starttime 2019-04-18 21:07:13 +0000 UTC","knative.dev/controller":"taskrun-controller"}
I0418 21:17:16.128] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:537","msg":"TIMEOUT check if nil run-giraffe-0","knative.dev/controller":"taskrun-controller"}                                
I0418 21:17:16.128] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:541","msg":"TIMEOUT not nil run-giraffe-0 timeout 30s runtime 29.98450907s","knative.dev/controller":"taskrun-controller"}
```

As far as the reconcile loop is concerned, the timeout is `30s` but the runtime is `29.98450907s` - right on the edge!

And the same thing for `run-giraffe-6`:

```
I0418 21:17:16.033] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:541","msg":"TIMEOUT not nil run-giraffe-6 timeout 30s runtime 29.843298838s","knative.dev/controller":"taskrun-controller"}
```

So ∴ the theory is that the timeout handler and the reconcile loop are slightly out of sync in their logic to determine when a timeout has occurred, i.e. `<-time.After(timeout)` is happening, but [the reonciler's logic](https://github.com/tektoncd/pipeline/blob/984049e315afc7dd5c381a87c3afe199906a18b0/pkg/reconciler/v1alpha1/taskrun/taskrun.go#L538) is slightly different.

Anyway I don't know what the fix is haha - at a glance they _SEEM_ to be doing the same thing. But the key imo is in making sure that exactly the same logic is being applied in both places OR in moving the logic to _handle_ the timeout case into the timeout handler itself (instead of triggering an entire reconcile).

Whew! (maybe :crossed_fingers::crossed_fingers: :crossed_fingers: )

![cat dancing](https://66.media.tumblr.com/tumblr_lv31lrvuUZ1r6t9aao1_500.gif)

We need to:
- [ ] Figure out a fix for this (haha)
- [ ] Add more logs to track how timeouts are being applied (for any future debugging)

**Comment by bobcatfish** [comment]
Can see similar stuff happening in https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/771/pull-tekton-pipeline-integration-tests/1118995157169999873/build-log.txt as well (with "improved" logs):

```
I0418 22:09:34.153] {"level":"info","logger":"controller.taskrun-controller","caller":"controller/controller.go:288","msg":"Reconcile succeeded. Time taken: 47.002Âµs.","knative.dev/controller":"taskrun-controller","knative.dev/key":"arendelle-vs2m2/run-giraffe-0"}
I0418 22:09:34.153] {"level":"info","logger":"controller.taskrun-controller","caller":"reconciler/timeout_handler.go:203","msg":"TIMEOUT HANDLER timed out for run-giraffe-5","knative.dev/controller":"taskrun-controller"}
I0418 22:09:34.153] {"level":"info","logger":"controller.taskrun-controller","caller":"reconciler/timeout_handler.go:200","msg":"TIMEOUT HANDLER finished for run-giraffe-5","knative.dev/controller":"taskrun-controller"}
I0418 22:09:34.154] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:248","msg":"TIMEOUT RECONCILER about to check if timeout for run-giraffe-5","knative.dev/controller":"taskrun-controller"}
I0418 22:09:34.154] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:531","msg":"TIMEOUT RECONCILER checking timeout for run-giraffe-5, starttime 2019-04-18 21:59:31 +0000 UTC","knative.dev/controller":"taskrun-controller"}
I0418 22:09:34.154] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:537","msg":"TIMEOUT RECONCILER check if nil run-giraffe-5","knative.dev/controller":"taskrun-controller"}
I0418 22:09:34.154] {"level":"info","logger":"controller.taskrun-controller","caller":"taskrun/taskrun.go:541","msg":"TIMEOUT RECONCILER not nil run-giraffe-5 timeout 30s runtime 29.943341324s","knative.dev/controller":"taskrun-controller"}
```

`timeout 30s runtime 29.943341324s` DUN DUN DUUUUN

**Comment by vdemeester** [comment]
`TestTaskRunTimeout` is back at being flaky...
/cc @tektoncd/core-maintainers

**Comment by vdemeester** [comment]
Alright, figured it out I think

```
  The Race Condition:

  1. TaskRun is created with a 1-second timeout
  2. The timeout check (line 186) happens before the reconcile() method is called
  3. failTaskRun() is called, which calls terminateStepsInPod()
  4. But tr.Status.Steps may be empty because pod status hasn't been fetched yet in this reconciliation cycle
  5. The test expects all steps to have Terminated status, but they're missing

  The Fix:

  We need to ensure step statuses are populated from the pod before calling failTaskRun(). 
```
